README.md
exfat-nofuse
============

Linux non-fuse read/write kernel driver for the exFAT, FAT12, FAT16 and vfat (FAT32) file systems.<br />
Originally ported from Android kernel v3.0.


Kudos to ksv1986 for the mutex patch!<br />
Thanks to JackNorris for being awesome and providing the clear_inode() patch.<br />
<br />
Big thanks to lqs for completing the driver!<br />
Big thanks to benpicco for fixing 3.11.y compatibility!<br />

Special thanks to github user AndreiLux for spreading the word about the leak!<br />


Installing as a stand-alone module:
====================================

    make
    sudo make install

To load the driver manually, run this as root:

    modprobe exfat

You may also specify custom toolchains by using CROSS_COMPILE flag, in my case:
>CROSS_COMPILE=../dorimanx-SG2-I9100-Kernel/android-toolchain/bin/arm-eabi-

Installing as a part of the kernel:
======================================

Let's take [linux] as the path to your kernel source dir...

	cd [linux]
	cp -rvf exfat-nofuse [linux]/fs/exfat

edit [linux]/fs/Kconfig
```
 menu "DOS/FAT/NT Filesystems"
  
  source "fs/fat/Kconfig"
 +source "fs/exfat/Kconfig"
  source "fs/ntfs/Kconfig"
  
 endmenu  
```
  


edit [linux]/fs/Makefile
```
  obj-$(CONFIG_FAT_FS)    += fat/
 +obj-$(CONFIG_EXFAT_FS)  += exfat/
  obj-$(CONFIG_BFS_FS)    += bfs/
```

	cd [linux]
	make menuconfig

Go to:
> File systems > DOS/FAT/NT
>   check exfat as MODULE (M) <br/>
>   (437) Default codepage for exFAT <br/>
>   (utf8) Default iocharset for exFAT <br/>

> ESC to main menu <br/>
> Save an Alternate Configuration File <br/>
> ESC ESC <br/>

build your kernel

Have fun.

Free Software for the Free Minds!
=================================


debian/changelog
exfat-dkms (1.2.24+20190524) unstable; urgency=medium

  * fix In the 5.0 kernel, only the mount namespace code should use the MS_*
     macos. Filesystems should use the SB_* ones.

 -- jimmy <jimmy@jimmy-alpha>  Fri, 22 Mar 2019 15:17:00 +0800

exfat-nofuse (1.2.24+20181106) unstable; urgency=medium

  * fix kernel 4.18 compile error

 -- jimmy <jimmy@jimmy-alpha>  Tue, 06 Nov 2018 09:13:31 +0800

exfat-nofuse (1.2.24+20180409) unstable; urgency=medium

  * fix kernel 4.16 compile error

 -- jimmy <jimmy@jimmy-alpha>  Mon, 09 Apr 2018 18:08:08 +0800

exfat-nofuse (1.2.24+20170905) unstable; urgency=medium

  * Support exfat capacity more than 2TB
  * Stop using CURRENT_TIME_SEC
  *    This is going away in 4.12 and is not y2038 safe.
  *    Use current_time(inode) and ktime_get_real_ts where appropriate.

 -- jimmy <jimmy@jimmy-alpha>  Tue, 05 Sep 2017 10:26:41 +0800

exfat-nofuse (1.2.24+20170617) unstable; urgency=medium

  * import samsung 1.2.24

 -- coolshou <coolshou2060@gmail.com>  Sat, 17 Jun 2017 11:07:30 +0800

exfat-nofuse (1.2.9+20170608) unstable; urgency=medium

  * fix for kernel 4.11

 -- coolshou <coolshou2060@gmail.com>  Thu, 08 Jun 2017 11:07:30 +0800

exfat-nofuse (1.2.9+20170405) unstable; urgency=medium

  * fix for kernel 4.9, 4.10

 -- coolshou <coolshou2060@gmail.com>  Wed, 05 Apr 2017 21:41:55 +0800

exfat-nofuse (1.2.9+20161213) unstable; urgency=medium

  [ coolshou ]
  * add support kernel 4.7

  [ jimmy ]
  * support kernel 4.8

 -- coolshou <coolshou@coolshou2060@gmail.com>  Tue, 13 Dec 2016 09:01:32 +0800

exfat-nofuse (1.2.9+20160316-0) unstable; urgency=medium

  * New upstream snapshot (a8d79fc258336dfe71a07e6c3e0f434a630b9e37)

 -- Eugene San (eugenesan) <eugenesan@gmail.com>  Sat, 02 Apr 2016 12:58:24 -0400

exfat-nofuse (1.2.9+20151110-0) unstable; urgency=medium

  * New upstream snapshot (593e1cef6219a932b281ee486fa68b252182c1f8)

 -- Eugene San (eugenesan) <eugenesan@gmail.com>  Thu, 12 Nov 2015 16:59:38 -0500

exfat-nofuse (1.2.9+20150620-0) unstable; urgency=medium

  * New upstream snapshot

 -- Eugene San (eugenesan) <eugenesan@gmail.com>  Sat, 25 Jul 2015 09:48:05 -0400

exfat-nofuse (1.2.9+20150125-0) unstable; urgency=medium

  * New upstream snapshot

 -- Eugene San (eugenesan) <eugenesan@gmail.com>  Mon, 16 Feb 2015 07:24:51 +0200

exfat-nofuse (1.2.9+20141015-0) unstable; urgency=medium

  * New upstream snapshot
  * Rename package after upstream

 -- Eugene San (eugenesan) <eugenesan@gmail.com>  Sun, 16 Nov 2014 10:01:04 +0200

exfat (1.2.9+20140805-0) unstable; urgency=medium

  * New upstream snapshot

 -- Eugene San (eugenesan) <eugenesan@gmail.com>  Sun, 05 Oct 2014 11:15:56 +0300

exfat (1.2.8+20140413-0) unstable; urgency=low

  * New upstream snapshot

 -- Eugene San (eugenesan) <eugenesan@gmail.com>  Sat, 19 Apr 2014 19:36:19 +0300

exfat (1.2.7-1) unstable; urgency=low

  * Initial release

 -- Eugene San (eugenesan) <eugenesan@gmail.com>  Sun, 19 Jan 2014 16:01:49 +0200


exfat_oal.c
/* Some of the source code in this file came from "linux/fs/fat/misc.c".  */
/*
 *  linux/fs/fat/misc.c
 *
 *  Written 1992,1993 by Werner Almesberger
 *  22/11/2000 - Fixed fat_date_unix2dos for dates earlier than 01/01/1980
 *         and date_dos2unix for date==0 by Igor Zhbanov(bsg@uniyar.ac.ru)
 */

/*
 *  Copyright (C) 2012-2013 Samsung Electronics Co., Ltd.
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <linux/semaphore.h>
#include <linux/time.h>

#include "exfat_config.h"
#include "exfat_global.h"
#include "exfat_api.h"
#include "exfat_oal.h"

/*======================================================================*/
/*                                                                      */
/*            SEMAPHORE FUNCTIONS                                       */
/*                                                                      */
/*======================================================================*/

#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,36)
DECLARE_MUTEX(z_sem);
#else
DEFINE_SEMAPHORE(z_sem);
#endif

INT32 sm_init(struct semaphore *sm)
{
	sema_init(sm, 1);
	return(0);
} /* end of sm_init */

INT32 sm_P(struct semaphore *sm)
{
	down(sm);
	return(0);
} /* end of sm_P */

void sm_V(struct semaphore *sm)
{
	up(sm);
} /* end of sm_V */


/*======================================================================*/
/*                                                                      */
/*            REAL-TIME CLOCK FUNCTIONS                                 */
/*                                                                      */
/*======================================================================*/

extern struct timezone sys_tz;

/*
 * The epoch of FAT timestamp is 1980.
 *     :  bits  : value
 * date:  0 -  4: day    (1 -  31)
 * date:  5 -  8: month  (1 -  12)
 * date:  9 - 15: year   (0 - 127) from 1980
 * time:  0 -  4: sec    (0 -  29) 2sec counts
 * time:  5 - 10: min    (0 -  59)
 * time: 11 - 15: hour   (0 -  23)
 */
#define UNIX_SECS_1980   315532800L

#if BITS_PER_LONG == 64
#define UNIX_SECS_2108   4354819200L
#endif
/* days between 1.1.70 and 1.1.80 (2 leap days) */
#define DAYS_DELTA_DECADE	(365 * 10 + 2)
/* 120 (2100 - 1980) isn't leap year */
#define NO_LEAP_YEAR_2100    (120)
#define IS_LEAP_YEAR(y)  (!((y) & 3) && (y) != NO_LEAP_YEAR_2100)

#define SECS_PER_MIN     (60)
#define SECS_PER_HOUR    (60 * SECS_PER_MIN)
#define SECS_PER_DAY     (24 * SECS_PER_HOUR)

#define MAKE_LEAP_YEAR(leap_year, year)                         \
	do {                                                    \
		if (unlikely(year > NO_LEAP_YEAR_2100))         \
			leap_year = ((year + 3) / 4) - 1;       \
		else                                            \
			leap_year = ((year + 3) / 4);           \
	} while(0)

/* Linear day numbers of the respective 1sts in non-leap years. */
static time_t accum_days_in_year[] = {
	/* Jan  Feb  Mar  Apr  May  Jun  Jul  Aug  Sep  Oct  Nov  Dec */
	0,   0,  31,  59,  90, 120, 151, 181, 212, 243, 273, 304, 334, 0, 0, 0,
};

TIMESTAMP_T *tm_current(TIMESTAMP_T *tp, UINT8 tz_utc)
{
	struct timespec ts;
	time_t second, day, leap_day, month, year;

#if LINUX_VERSION_CODE < KERNEL_VERSION(4,8,0)
	ts = CURRENT_TIME_SEC;
#else
	ktime_get_real_ts(&ts);
#endif

	second = ts.tv_sec;
	if (!tz_utc)
		second -= sys_tz.tz_minuteswest * SECS_PER_MIN;

	/* Jan 1 GMT 00:00:00 1980. But what about another time zone? */
	if (second < UNIX_SECS_1980) {
		tp->sec  = 0;
		tp->min  = 0;
		tp->hour = 0;
		tp->day  = 1;
		tp->mon  = 1;
		tp->year = 0;
		return(tp);
	}
#if BITS_PER_LONG == 64
	if (second >= UNIX_SECS_2108) {
		tp->sec  = 59;
		tp->min  = 59;
		tp->hour = 23;
		tp->day  = 31;
		tp->mon  = 12;
		tp->year = 127;
		return(tp);
	}
#endif

	day = second / SECS_PER_DAY - DAYS_DELTA_DECADE;
	year = day / 365;

	MAKE_LEAP_YEAR(leap_day, year);
	if (year * 365 + leap_day > day)
		year--;

	MAKE_LEAP_YEAR(leap_day, year);

	day -= year * 365 + leap_day;

	if (IS_LEAP_YEAR(year) && day == accum_days_in_year[3]) {
		month = 2;
	} else {
		if (IS_LEAP_YEAR(year) && day > accum_days_in_year[3])
			day--;
		for (month = 1; month < 12; month++) {
			if (accum_days_in_year[month + 1] > day)
				break;
		}
	}
	day -= accum_days_in_year[month];

	tp->sec  = second % SECS_PER_MIN;
	tp->min  = (second / SECS_PER_MIN) % 60;
	tp->hour = (second / SECS_PER_HOUR) % 24;
	tp->day  = day + 1;
	tp->mon  = month;
	tp->year = year;

	return(tp);
} /* end of tm_current */

/* end of exfat_oal.c */
